src/tunnel.cc

```diff
@@ -971,20 +972,21 @@ tunnelStart(ClientHttpRequest * http)
 }
 
 void
-TunnelStateData::connectToPeer()
+TunnelStateData::connectToPeer(const Comm::ConnectionPointer &conn)
 {
-    if (CachePeer *p = server.conn->getPeer()) {
+    if (CachePeer *p = conn->getPeer()) {
         if (p->secure.encryptTransport) {
             AsyncCall::Pointer callback = asyncCall(5,4,
                                                     "TunnelStateData::ConnectedToPeer",
                                                     MyAnswerDialer(&TunnelStateData::connectedToPeer, this));
-            auto *connector = new Security::BlindPeerConnector(request, server.conn, callback, al);
+            auto *connector = new Security::BlindPeerConnector(request, conn, callback, al);
             AsyncJob::Start(connector); // will call our callback
             return;
         }
     }
 
     Security::EncryptorAnswer nil;
+    nil.conn = conn;
```

Comment:
This nil "dereference" does not look good. Please split TunnelStateData::connectedToPeer() into two methods:
* noteSecurityPeerConnectorAnswer() which either handles errors or calls connectedToPeer()
* connectedToPeer() which receives a usable to-peer TCP connection (already secured with TLS where needed).

This connectToPeer() method will then just call `connectedToPeer(conn)` instead of using a nil answer trick.

Reply:
commit 6bcaba103a